File: activetogether-backend/ActiveTogether.Services/Database/Entities/Activity.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using ActiveTogether.Model.Enums;

namespace ActiveTogether.Services.Database.Entities
{
    public class Activity
    {
        [Key]
        public int Id { get; set; }

        [Required, MaxLength(150)]
        public string Name { get; set; } = string.Empty;

        [Required, MaxLength(1000)]
        public string Description { get; set; } = string.Empty;

        [ForeignKey(nameof(Category))]
        public int CategoryId { get; set; }
        public Category? Category { get; set; }

        [ForeignKey(nameof(ActivityType))]
        public int ActivityTypeId { get; set; }
        public ActivityType? ActivityType { get; set; }

        [ForeignKey(nameof(Location))]
        public int LocationId { get; set; }
        public Location? Location { get; set; }

        [ForeignKey(nameof(Organizer))]
        public int OrganizerId { get; set; }
        public User? Organizer { get; set; }

        public DateTime DateTime { get; set; }

        public int Capacity { get; set; }

        public bool IsFree { get; set; } = true;

        public decimal? Price { get; set; }

        public string? ImageUrl { get; set; }

        public ActivityStatus Status { get; set; } = ActivityStatus.Draft;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }

        public ICollection<Reservation> Reservations { get; set; } = new List<Reservation>();
    }
}

File: activetogether-frontend/activetogether_mobile/lib/models/activity.dart
class Activity {
  final int id;
  final String name;
  final String description;
  final int categoryId;
  final String categoryName;
  final int activityTypeId;
  final String activityTypeName;
  final int locationId;
  final String locationName;
  final String locationAddress;
  final double locationLatitude;
  final double locationLongitude;
  final int organizerId;
  final String organizerName;
  final DateTime dateTime;
  final int capacity;
  final int reservedCount;
  final bool isFree;
  final double? price;
  final String? imageUrl;
  final String status;
  final double? averageRating;
  final int ratingCount;

  Activity({
    required this.id,
    required this.name,
    required this.description,
    required this.categoryId,
    required this.categoryName,
    required this.activityTypeId,
    required this.activityTypeName,
    required this.locationId,
    required this.locationName,
    required this.locationAddress,
    this.locationLatitude = 0,
    this.locationLongitude = 0,
    required this.organizerId,
    required this.organizerName,
    required this.dateTime,
    required this.capacity,
    required this.reservedCount,
    required this.isFree,
    this.price,
    this.imageUrl,
    required this.status,
    this.averageRating,
    required this.ratingCount,
  });

  int get spotsLeft => capacity - reservedCount;
  double get fillRatio => capacity == 0 ? 0 : reservedCount / capacity;

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'] as int,
      name: json['name'] as String,
      description: json['description'] as String? ?? '',
      categoryId: json['categoryId'] as int,
      categoryName: json['categoryName'] as String? ?? '',
      activityTypeId: json['activityTypeId'] as int,
      activityTypeName: json['activityTypeName'] as String? ?? '',
      locationId: json['locationId'] as int,
      locationName: json['locationName'] as String? ?? '',
      locationAddress: json['locationAddress'] as String? ?? '',
      locationLatitude: (json['locationLatitude'] as num?)?.toDouble() ?? 0,
      locationLongitude: (json['locationLongitude'] as num?)?.toDouble() ?? 0,
      organizerId: json['organizerId'] as int,
      organizerName: json['organizerName'] as String? ?? '',
      dateTime: DateTime.parse(json['dateTime'] as String),
      capacity: json['capacity'] as int,
      reservedCount: json['reservedCount'] as int,
      isFree: json['isFree'] as bool,
      price: (json['price'] as num?)?.toDouble(),
      imageUrl: json['imageUrl'] as String?,
      status: json['status'] as String,
      averageRating: (json['averageRating'] as num?)?.toDouble(),
      ratingCount: json['ratingCount'] as int? ?? 0,
    );
  }
}
